File: macos/Sources/ElevateCore/Providers/EntraDirectoryProvider.swift
import Foundation

public struct EntraDirectoryProvider: PIMProvider {
    public let kind: RoleScopeKind = .entraDirectory
    public let scopes = GraphScopes.all
    let transport: GraphTransport

    public init(http: any HTTPClient, tokens: any TokenProviding) {
        transport = GraphTransport(http: http, tokens: tokens)
    }

    // MARK: Wire models

    struct RoleDefinitionRef: Decodable { let id: String; let displayName: String? }
    struct Schedule: Decodable {
        let id: String
        let roleDefinitionId: String
        let directoryScopeId: String?
        let assignmentType: String?
        let startDateTime: Date?
        let endDateTime: Date?
        let roleDefinition: RoleDefinitionRef?
        let memberType: String?
    }
    struct ScheduleRequest: Decodable {
        let id: String
        let status: String
        let roleDefinitionId: String
        let directoryScopeId: String?
        let createdDateTime: Date?
        let scheduleInfo: ScheduleInfo?
    }
    struct Collection<T: Decodable>: Decodable { let value: [T] }

    // MARK: Reads

    public func eligibleRoles(identity: Identity, tenant: TenantContext) async throws -> [EligibleRole] {
        let r = try await transport.get(identity: identity, tenantId: tenant.tenantId,
                                        url: try transport.graphURL("/roleManagement/directory/roleEligibilitySchedules/filterByCurrentUser(on='principal')?$expand=roleDefinition"),
                                        scopes: scopes)
        let items = try GraphJSON.decoder.decode(Collection<Schedule>.self, from: r.body).value
        var seen = Set<RoleScope>()
        var roles: [EligibleRole] = []
        for s in items {
            let scope = RoleScope.entraDirectory(roleDefinitionId: s.roleDefinitionId, directoryScopeId: s.directoryScopeId ?? "/")
            guard seen.insert(scope).inserted else { continue }
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId, scope: scope)
            let viaGroup: String? = s.memberType?.caseInsensitiveCompare("Group") == .orderedSame ? "group" : nil
            roles.append(EligibleRole(key: key, displayName: s.roleDefinition?.displayName ?? s.roleDefinitionId,
                                      source: .discovered, policy: .manualDefault, viaGroup: viaGroup))
        }
        return roles.sorted { $0.displayName < $1.displayName }
    }

    public func activeAssignments(identity: Identity, tenant: TenantContext) async throws -> [ActiveAssignment] {
        let instances = try await transport.get(identity: identity, tenantId: tenant.tenantId,
                                                url: try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleInstances/filterByCurrentUser(on='principal')?$expand=roleDefinition"),
                                                scopes: scopes)
        // Widened past PendingApproval so a booked-ahead request, which the service has already
        // turned into a schedule, is the source for the `.scheduled` rows below.
        let requests = try await transport.get(identity: identity, tenantId: tenant.tenantId,
                                               url: try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleRequests/filterByCurrentUser(on='principal')?$filter=status eq 'PendingApproval' or status eq 'ScheduleCreated' or status eq 'Provisioned'"),
                                               scopes: scopes)
        let activated = try GraphJSON.decoder.decode(Collection<Schedule>.self, from: instances.body).value
            .filter { $0.assignmentType == "Activated" }
        let all = try GraphJSON.decoder.decode(Collection<ScheduleRequest>.self, from: requests.body).value
        let pending = all.filter { $0.status == "PendingApproval" }

        var result: [RoleKey: ActiveAssignment] = [:]
        for s in activated {
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId,
                              scope: .entraDirectory(roleDefinitionId: s.roleDefinitionId, directoryScopeId: s.directoryScopeId ?? "/"))
            result[key] = ActiveAssignment(roleKey: key, assignmentId: s.id, startDateTime: s.startDateTime ?? .now,
                                           endDateTime: s.endDateTime, status: .active)
        }
        for p in pending {
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId,
                              scope: .entraDirectory(roleDefinitionId: p.roleDefinitionId, directoryScopeId: p.directoryScopeId ?? "/"))
            guard result[key] == nil else { continue }
            result[key] = ActiveAssignment(roleKey: key, assignmentId: p.id,
                                           startDateTime: p.scheduleInfo?.startDateTime ?? p.createdDateTime ?? .now,
                                           endDateTime: nil, status: .pendingApproval)
        }
        for u in all where !ScheduledStart.isSettledOrPending(u.status) {
            guard let start = u.scheduleInfo?.startDateTime, ScheduledStart.isFuture(start) else { continue }
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId,
                              scope: .entraDirectory(roleDefinitionId: u.roleDefinitionId, directoryScopeId: u.directoryScopeId ?? "/"))
            guard result[key] == nil else { continue }
            let end = ScheduledStart.end(explicit: u.scheduleInfo?.expiration?.endDateTime,
                                         duration: u.scheduleInfo?.expiration?.duration, start: start, fallback: nil)
            result[key] = ActiveAssignment(roleKey: key, assignmentId: u.id, startDateTime: start,
                                           endDateTime: end, status: .scheduled)
        }
        return Array(result.values)
    }

    // MARK: Policy

    struct Policy: Decodable { let id: String; let rules: [PolicyRules.Rule]? }
    struct PolicyAssignment: Decodable { let id: String; let roleDefinitionId: String?; let policy: Policy? }

    public func policy(for role: EligibleRole, identity: Identity) async throws -> RolePolicy {
        guard case .entraDirectory(let roleDefinitionId, _) = role.key.scope else { throw PIMError.notEligible }
        let filter = "scopeId eq '/' and scopeType eq 'DirectoryRole' and roleDefinitionId eq '\(roleDefinitionId)'"
        let encoded = filter.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? filter
        let r = try await transport.get(identity: identity, tenantId: role.key.tenantId,
                                        url: try transport.graphURL("/policies/roleManagementPolicyAssignments?$filter=\(encoded)&$expand=policy($expand=rules)"),
                                        scopes: scopes)
        let assignments = try GraphJSON.decoder.decode(Collection<PolicyAssignment>.self, from: r.body).value
        guard let rules = assignments.first?.policy?.rules else { return .manualDefault }
        return PolicyRules.apply(rules)
    }

    // MARK: Activate / deactivate

    struct Me: Decodable { let id: String }

    func principalId(identity: Identity, tenantId: String) async throws -> String {
        let r = try await transport.get(identity: identity, tenantId: tenantId, url: try transport.graphURL("/me?$select=id"), scopes: scopes)
        return try GraphJSON.decoder.decode(Me.self, from: r.body).id
    }

    public func activate(_ request: ActivationRequest, identity: Identity) async throws -> ActiveAssignment {
        guard case .entraDirectory(let roleDefinitionId, let directoryScopeId) = request.roleKey.scope else { throw PIMError.notEligible }
        let principal = try await principalId(identity: identity, tenantId: request.roleKey.tenantId)
        var body: [String: Any] = [
            "action": "selfActivate",
            "principalId": principal,
            "roleDefinitionId": roleDefinitionId,
            "directoryScopeId": directoryScopeId,
            "justification": request.justification,
            "scheduleInfo": [
                "startDateTime": GraphJSON.encoderDateString(request.startDateTime ?? .now),
                "expiration": ["type": "afterDuration", "duration": ISO8601Duration.format(request.duration)],
            ],
        ]
        if let t = request.ticket {
            body["ticketInfo"] = ["ticketNumber": t.number, "ticketSystem": t.system]
        }
        let r = try await transport.post(identity: identity, tenantId: request.roleKey.tenantId,
                                         url: try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleRequests"),
                                         scopes: scopes, body: try JSONSerialization.data(withJSONObject: body))
        let created = try GraphJSON.decoder.decode(ScheduleRequest.self, from: r.body)
        let start = ScheduledStart.effective(response: created.scheduleInfo?.startDateTime, requested: request.startDateTime)
        let end = ScheduledStart.end(explicit: created.scheduleInfo?.expiration?.endDateTime,
                                     duration: created.scheduleInfo?.expiration?.duration, start: start, fallback: request.duration)
        let reported: ActiveAssignment.Status = switch created.status {
        case "PendingApproval", "PendingAdminDecision": .pendingApproval
        case "PendingProvisioning", "PendingScheduleCreation", "ScheduleCreated": .pendingProvisioning
        case "Denied", "Failed", "Canceled", "Revoked": .failed(created.status)
        default: .active
        }
        // A future start only masks an outcome that would otherwise read as active; pending/failed still win.
        let status: ActiveAssignment.Status = (reported == .active && ScheduledStart.isFuture(start)) ? .scheduled : reported
        return ActiveAssignment(roleKey: request.roleKey, assignmentId: created.id, startDateTime: start,
                                endDateTime: status == .active || status == .scheduled ? end : nil, status: status)
    }

    public func deactivate(_ assignment: ActiveAssignment, identity: Identity) async throws {
        guard case .entraDirectory(let roleDefinitionId, let directoryScopeId) = assignment.roleKey.scope else { throw PIMError.notEligible }
        let principal = try await principalId(identity: identity, tenantId: assignment.roleKey.tenantId)
        let body: [String: Any] = [
            "action": "selfDeactivate",
            "principalId": principal,
            "roleDefinitionId": roleDefinitionId,
            "directoryScopeId": directoryScopeId,
        ]
        _ = try await transport.post(identity: identity, tenantId: assignment.roleKey.tenantId,
                                     url: try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleRequests"),
                                     scopes: scopes, body: try JSONSerialization.data(withJSONObject: body))
    }

    /// Withdraws a request still awaiting approval. Graph answers 204 with no body.
    public func cancelPendingRequest(_ assignment: ActiveAssignment, identity: Identity) async throws {
        guard let requestId = assignment.assignmentId else { throw PIMError.notEligible }
        _ = try await transport.post(identity: identity, tenantId: assignment.roleKey.tenantId,
                                     url: try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleRequests/\(requestId)/cancel"),
                                     scopes: scopes, body: Data())
    }
}

File: windows/src/Elevate.Core/Providers/EntraDirectoryProvider.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;
using Elevate.Core.Support;

namespace Elevate.Core.Providers;

/// <summary>PIM for Entra directory roles through Microsoft Graph. Port of the Swift <c>EntraDirectoryProvider</c>.</summary>
public sealed class EntraDirectoryProvider : IPimProvider
{
    private readonly GraphTransport _transport;

    public EntraDirectoryProvider(IHttpClient http, ITokenProvider tokens)
        => _transport = new GraphTransport(http, tokens);

    public RoleScopeKind Kind => RoleScopeKind.EntraDirectory;

    public IReadOnlyList<string> Scopes => Auth.Scopes.GraphAll;

    // MARK: Wire models

    private sealed record RoleDefinitionRef(string Id, string? DisplayName);

    private sealed record Schedule(
        string Id,
        string RoleDefinitionId,
        string? DirectoryScopeId,
        string? AssignmentType,
        DateTimeOffset? StartDateTime,
        DateTimeOffset? EndDateTime,
        RoleDefinitionRef? RoleDefinition,
        string? MemberType);

    private sealed record ScheduleRequest(
        string Id,
        string Status,
        string RoleDefinitionId,
        string? DirectoryScopeId,
        DateTimeOffset? CreatedDateTime,
        ScheduleInfo? ScheduleInfo);

    private sealed record Collection<T>(IReadOnlyList<T> Value);

    private sealed record Me(string Id);

    private static T Decode<T>(HttpResponseData response) =>
        JsonSerializer.Deserialize<T>(response.Body, GraphJson.Options)
        ?? throw new PimException(PimErrorKind.Unexpected, "Empty response body");

    // MARK: Reads

    public async Task<IReadOnlyList<EligibleRole>> EligibleRolesAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var items = await _transport.ListAllAsync<Schedule>(
            identity, tenant.TenantId,
            _transport.GraphUrl("/roleManagement/directory/roleEligibilitySchedules/filterByCurrentUser(on='principal')?$expand=roleDefinition"),
            Scopes, ct).ConfigureAwait(false);
        var seen = new HashSet<RoleScope>();
        var roles = new List<EligibleRole>();
        foreach (var s in items)
        {
            var scope = new EntraDirectoryScope(s.RoleDefinitionId, s.DirectoryScopeId ?? "/");
            if (!seen.Add(scope))
            {
                continue;
            }

            var key = new RoleKey(identity.Id, tenant.TenantId, scope);
            var viaGroup = string.Equals(s.MemberType, "Group", StringComparison.OrdinalIgnoreCase) ? "group" : null;
            roles.Add(new EligibleRole(
                key, s.RoleDefinition?.DisplayName ?? s.RoleDefinitionId,
                RoleSource.Discovered, RolePolicy.ManualDefault, ViaGroup: viaGroup));
        }

        return [.. roles.OrderBy(r => r.DisplayName, StringComparer.Ordinal)];
    }

    public async Task<IReadOnlyList<ActiveAssignment>> ActiveAssignmentsAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var instances = await _transport.ListAllAsync<Schedule>(
            identity, tenant.TenantId,
            _transport.GraphUrl("/roleManagement/directory/roleAssignmentScheduleInstances/filterByCurrentUser(on='principal')?$expand=roleDefinition"),
            Scopes, ct).ConfigureAwait(false);

        // Widened past PendingApproval so a booked-ahead request, which the service has already
        // turned into a schedule, is the source for the scheduled rows below.
        var all = await _transport.ListAllAsync<ScheduleRequest>(
            identity, tenant.TenantId,
            _transport.GraphUrl("/roleManagement/directory/roleAssignmentScheduleRequests/filterByCurrentUser(on='principal')?$filter=status eq 'PendingApproval' or status eq 'ScheduleCreated' or status eq 'Provisioned'"),
            Scopes, ct).ConfigureAwait(false);

        var activated = instances.Where(s => s.AssignmentType == "Activated");

        var result = new Dictionary<RoleKey, ActiveAssignment>();
        foreach (var s in activated)
        {
            var key = new RoleKey(identity.Id, tenant.TenantId, new EntraDirectoryScope(s.RoleDefinitionId, s.DirectoryScopeId ?? "/"));
            result[key] = new ActiveAssignment(
                key, s.Id, s.StartDateTime ?? DateTimeOffset.UtcNow, s.EndDateTime, AssignmentStatus.Active);
        }

        foreach (var p in all.Where(r => r.Status == "PendingApproval"))
        {
            var key = new RoleKey(identity.Id, tenant.TenantId, new EntraDirectoryScope(p.RoleDefinitionId, p.DirectoryScopeId ?? "/"));
            if (result.ContainsKey(key))
            {
                continue;
            }

            result[key] = new ActiveAssignment(
                key, p.Id,
                p.ScheduleInfo?.StartDateTime ?? p.CreatedDateTime ?? DateTimeOffset.UtcNow,
                null, AssignmentStatus.PendingApproval);
        }

        foreach (var u in all.Where(r => !ScheduleRules.IsSettledOrPending(r.Status)))
        {
            if (u.ScheduleInfo?.StartDateTime is not { } start || !ScheduleRules.IsFuture(start))
            {
                continue;
            }

            var key = new RoleKey(identity.Id, tenant.TenantId, new EntraDirectoryScope(u.RoleDefinitionId, u.DirectoryScopeId ?? "/"));
            if (result.ContainsKey(key))
            {
                continue;
            }

            var end = ScheduleRules.End(u.ScheduleInfo?.Expiration?.EndDateTime, u.ScheduleInfo?.Expiration?.Duration, start);
            result[key] = new ActiveAssignment(key, u.Id, start, end, AssignmentStatus.Scheduled);
        }

        return [.. result.Values];
    }

    // MARK: Policy

    private sealed record Policy(string Id, IReadOnlyList<PolicyRules.Rule>? Rules);

    private sealed record PolicyAssignment(string Id, string? RoleDefinitionId, Policy? Policy);

    public async Task<RolePolicy> PolicyAsync(EligibleRole role, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(role);

        if (role.Key.Scope is not EntraDirectoryScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var filter = $"scopeId eq '/' and scopeType eq 'DirectoryRole' and roleDefinitionId eq '{scope.RoleDefinitionId}'";
        var encoded = GraphTransport.PercentEncodeQuery(filter);
        var response = await _transport.GetAsync(
            identity, role.Key.TenantId,
            _transport.GraphUrl($"/policies/roleManagementPolicyAssignments?$filter={encoded}&$expand=policy($expand=rules)"),
            Scopes, ct).ConfigureAwait(false);

        var assignments = Decode<Collection<PolicyAssignment>>(response).Value;
        var rules = assignments.Count > 0 ? assignments[0].Policy?.Rules : null;
        return rules is null ? RolePolicy.ManualDefault : PolicyRules.Apply(rules);
    }

    // MARK: Activate / deactivate

    private async Task<string> PrincipalIdAsync(Identity identity, string tenantId, CancellationToken ct)
    {
        var response = await _transport.GetAsync(identity, tenantId, _transport.GraphUrl("/me?$select=id"), Scopes, ct)
            .ConfigureAwait(false);
        return Decode<Me>(response).Id;
    }

    public async Task<ActiveAssignment> ActivateAsync(
        ActivationRequest request, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        if (request.RoleKey.Scope is not EntraDirectoryScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var principal = await PrincipalIdAsync(identity, request.RoleKey.TenantId, ct).ConfigureAwait(false);
        var body = new JsonObject
        {
            ["action"] = "selfActivate",
            ["principalId"] = principal,
            ["roleDefinitionId"] = scope.RoleDefinitionId,
            ["directoryScopeId"] = scope.DirectoryScopeId,
            ["justification"] = request.Justification,
            ["scheduleInfo"] = new JsonObject
            {
                ["startDateTime"] = GraphJson.EncoderDateString(request.StartDateTime ?? DateTimeOffset.UtcNow),
                ["expiration"] = new JsonObject
                {
                    ["type"] = "afterDuration",
                    ["duration"] = Iso8601Duration.Format(request.Duration),
                },
            },
        };

        if (request.Ticket is { } ticket)
        {
            body["ticketInfo"] = new JsonObject
            {
                ["ticketNumber"] = ticket.Number,
                ["ticketSystem"] = ticket.System,
            };
        }

        var response = await _transport.PostAsync(
            identity, request.RoleKey.TenantId,
            _transport.GraphUrl("/roleManagement/directory/roleAssignmentScheduleRequests"),
            Scopes, System.Text.Encoding.UTF8.GetBytes(body.ToJsonString()), ct).ConfigureAwait(false);

        var created = Decode<ScheduleRequest>(response);
        var start = ScheduleRules.Effective(created.ScheduleInfo?.StartDateTime, request.StartDateTime);
        var end = ScheduleRules.End(
            created.ScheduleInfo?.Expiration?.EndDateTime, created.ScheduleInfo?.Expiration?.Duration,
            start, request.Duration);

        var (status, reportedEnd) = GraphSchedule.Settle(created.Status, start, end);
        return new ActiveAssignment(request.RoleKey, created.Id, start, reportedEnd, status);
    }

    public async Task DeactivateAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);

        if (assignment.RoleKey.Scope is not EntraDirectoryScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var principal = await PrincipalIdAsync(identity, assignment.RoleKey.TenantId, ct).ConfigureAwait(false);
        var body = new JsonObject
        {
            ["action"] = "selfDeactivate",
            ["principalId"] = principal,
            ["roleDefinitionId"] = scope.RoleDefinitionId,
            ["directoryScopeId"] = scope.DirectoryScopeId,
        };

        await _transport.PostAsync(
            identity, assignment.RoleKey.TenantId,
            _transport.GraphUrl("/roleManagement/directory/roleAssignmentScheduleRequests"),
            Scopes, System.Text.Encoding.UTF8.GetBytes(body.ToJsonString()), ct).ConfigureAwait(false);
    }

    /// <summary>Withdraws a request still awaiting approval. Graph answers 204 with no body.</summary>
    public async Task CancelPendingRequestAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);

        if (assignment.AssignmentId is not { } requestId)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        await _transport.PostAsync(
            identity, assignment.RoleKey.TenantId,
            _transport.GraphUrl($"/roleManagement/directory/roleAssignmentScheduleRequests/{requestId}/cancel"),
            Scopes, [], ct).ConfigureAwait(false);
    }
}
